sonic-xcvrd/scripts/xcvrd

```diff
@@ -613,6 +621,42 @@ class sfp_state_update_task:
         self.task_process = None
         self.task_stopping_event = multiprocessing.Event()
 
+    def _retry_for_system_ready(self):
+        retry = 0
+        while retry < RETRY_TIMES_FOR_SYSTEM_READY:
+            #time.sleep(RETRY_PERIOD_FOR_SYSTEM_READY_SECS)
+            time_start  = time.time()
+            rc, event_dict = platform_sfputil.get_transceiver_change_event(RETRY_PERIOD_FOR_SYSTEM_READY_MSECS)
```

Comment:
>platform_sfputil.get_transceiver_change_event [](start = 29, length = 45)

You can call platform_sfputil.get_transceiver_change_event function only once, and implement the logic as a normal state machine.

I don't mean your implementation is wrong, state machine may be much readable and future proof.

Reply:
Hi Qi,
I admit that rewriting it in a state machine approach can make the code a little shorter by eliminating some duplicate code.
But I'm afraid it can also introduce negative effects: We have the logic to control retry timespan and times and introduce some local variables to control that. Rewriting it as a state machine will combine the retry control logic into the main procedure of xcvrd, which make the main function complicated.

Reply:
The benefit for state machine pattern is not shorter code, but a correct design and implementation.

For the design part, you need to consider any sequence of change events returned by vendor plugin code. For example, SYSTEM_NOT_READY followed by an inserted event. You need to draw the state machine on paper first to consider all the events at each state.

For the implementation part, you could design you state transition code in standalone function, and achieve 'control retry timespan and times' or 'local variables'.

---
In reply to: [310075723](https://github.com/Azure/sonic-platform-daemons/pull/39#discussion_r310075723) [](ancestors = 310075723)

Reply:
Hi Qi,
I understand the advantage of state machine. I just feel that the logic of xcvrd daemon is relatively simple and it seems that using state machine will require a bit more effort.

Reply:
I've also summarized the state machine which may reflect the logic. you can also check and compare.

states
- Initial state: INIT, before received system ready or a normal event
- Final state: EXIT
- other state: NORMAL, after has received system-ready or a normal event

events
- SYSTEM_NOT_READY
- SYSTEM_BECOME_READY
- NORMAL EVENT
  - sfp insertion/removal
  - timeout returned by sfputil.get_change_event with status = true
- SYSTEM_FAIL

State transmit:
1.	Current state: INIT
    - On receiving event SYSTEM_NOT_READY, handle by on_init_snr
        If retry < RETRY_TIMES_FOR_SYSTEM_READY : retry++
        else: reach max retry, exit
    - On receiving event SYSTEM_BECOME_READY, handle by on_init_sr
        Transmit state to NORMAL
    - On receiving event SYSTEM_FAIL, handle by on_anystate_sf
        Treat as a fatal error, exit
    - On receiving normal event like SFP modules insertion/removal (for the vendors who don’t implement SYSTEM_BECOME_READY), handle by on_anystate_normal
      Transmit state to NORMAL
      Handle the plugin/plugout event normally.
2.	Current state: NORMAL
    - On receiving normal event, handled by on_anystate_normal
        Handle the event normally.
        Don’t transmit state.
    - On receiving SYSTEM_BECOME_READY
        Just log the event and ignored
    - On receiving SYSTEM_NOT_READY, handle by on_normal_snr
        Treat as a fatal error, exit
    - On receiving SYSTEM_FAIL, handle by on_anystate_sf
        Treat as a fatal error, exit

State           |event               next state
INIT            |SYSTEM NOT READY    |INIT / EXIT
INIT            |SYSTEM READY        |NORMAL
ANYSTATE        |SYSTEM FAIL         |EXIT
ANYSTATE        |NORMAL EVENT        |NORMAL
NORMAL          |SYSTEM NOT READY    |EXIT
EXIT            -